### DISCUSSION TITLE
I made PyMOL run on web-browsers (WASM port)

### DISCUSSION THREAD

**Discussion by yakomaxa** [schrodinger/pymol-open-source#307] [Show and tell]
Dear PyMOL developers. 

First of all, thank you for maintaining this wonderful software. 

I'm a bioinformatician in Japan, and I have been trying to port PyMOL to Web-assembly (WASM). I've successfully ported it using Pyodide, and now my PyMOL (the internal GUI part) runs on modern browsers like Chrome and Firefox with the help of a bit of javascript code. Though it's still under development, I want to share this with the community. However, I'm hesitating to share because I'm unsure about two things: 

(1) Is it acceptable to distribute PyMOL wasm binary by myself, especially naming it "PyMOL-wasm" or something like that?
(2) How should I distribute the source code? I considered making a pull request for PyMOL main, but I guess my modifications are large and also out of the main branch's scope, so I'm hesitating about this, too. 

On (1) point, I sent an email to Schrodinger Inc. months ago for confirmations, but there are no replies. 

I created this discussion to ask you (developers here) what would be the best way to share this work. I hope you'll be interested in this work, and any suggestions and questions will be appreciated.

Best regards, 
Koya Sakuma @yakomaxa

**Comment by JarrettSJohnson** [comment]
Hi Koya, we look forward to seeing the progress of your work! Excited to see what you have. At this moment, feel free to distribute the binaries; `PyMOL-wasm` sounds fine.

As far as merging the project in, let's experiment with having it separate from the main repository at least for now. Depending on your implementation, if pymol-open-source itself has to be modified, perhaps keep it as a fork. But if it's unmodified, perhaps as a submodule is better? For projects similar we've had to this, it's inevitable for the main repo to also include some changes to accommodate the implementation for that project, and we'll usually be open to PRs for those as long as the changes are easily maintainable not intrusive.

**Reply by yakomaxa** [reply]
Hi Jarrett. Thank you for responding to this discussion and letting me distribute the binary! Making binary available allows me to open a simple website on the GitHub page for a demonstration. Please wait a moment for me to construct the website.  

As for source code, I now store the source code for the wasm port in my private repo, which can be disclosed at any time. I separated it from the forked repo to make it easy to make modifications when I started this project. With some effort, this source code (or modified parts) can be again merged to the fork of the latest original repo.  

In my implementation, pymol-open-source has to be modified. I've made (I believe) these modifications conditional using ifdef so that it will be compatible with the usual pymol build. However, this also needs extra packages, changes on setup.py, and so on for compilation. I'm considering how to deal with these parts now. Any suggestions on these issues will help. When we solve these issues, the source code can be safely imported into the forked repository, and I can create a pull request with it. Before it, I'll keep my version in my repo and try to merge it to the forked repo for ease of following updates in the main.

**Reply by yakomaxa** [reply]
Hi, I've opened this project's repo and demo page! The repo only contains binary and html/JavaScript at this moment. The demo page remains quite primitive and will be updated. 

Repository: 
https://github.com/yakomaxa/PyMOL-Wasm

Demo page:
https://yakomaxa.github.io/PyMOL-Wasm

I checked the interactions with the PyMOL window on my machine, which will be improved for other platforms.

It is worth noting that the current implementation cannot follow the changes in window size after launching the PyMOL process. I will also improve this in the future. 

I appreciate any comments on this project. Please post it to the repo's issues and discussions.

**Comment by galeo** [comment]
Hi @yakomaxa

Thank you for your work on “PyMOL-wasm”, which has made it possible for us to run PyMOL in web-browsers. You previously mentioned that the related code would be open-sourced, but that hasn’t happened yet. I’m also interested in this project, and recently tried building a new version based on the latest master branch of `pymol-open-source` following the approach you gave.

I haven’t done thorough testing—it just barely works, and there are many issues. Some of them are the ones you already mentioned in the `PyMOL-Wasm` issue list, such as yakomaxa/PyMOL-Wasm#28, unable to set transparency etc. There are also some new problems, like unable to set labels, schrodinger/pymol-open-source#447.

If people are interested, I can provide a `.whl` file. But as I mentioned, this version has many problems, and I don’t have development experience in this area. Without official support, I feel that some issues will be very difficult to resolve. So what I’d like to know is the official stance on this version and whether there are any plans to support its development—for example, whether an official PyMOL-wasm release might be made available in the future. These questions might be best answered by @JarrettSJohnson.

Thanks again for your efforts.

**Reply by yakomaxa** [reply]
Hi @galeo,

Sorry for the delayed response, I just noticed today that I was mentioned weeks ago.

I'm really interested in your work. And I'm impressed that you reproduced it from my really rough descriptions. I feel sorry that I haven't opened the code. 

I have some questions; especially, what version of PyMOL did you use as the base code? I made mine using 2.x, but I believe today's PyMOL is 3.x. If it's based on 3.x, what happened to the GUI stuff when you ported it to WASM?

I hacked PyMOL in a way that lacks proper git history (I mean I copied code, modified it, and pushed it to my private repo). I know this isn’t a good practice, and I wanted to restart clean from the official pymol-open-source repo before making it public. But I accidentally got really busy, and it’s been hard to find the time.

I believe the core C++ code of PyMOL didn’t change much between 2.x and 3.x, so I think I could redo my work for 3.x.

Sorry for being vague, but I’d really appreciate any feedback. Thanks a lot for your attention!

If you are not uncomfortable, I’m willing to discuss more about this, including how to handle this situation. 

Added (after posted): I believe pymol-open-source devs has no responsibility for my Wasm version. I will be so happy if we could integrate it into official codebase, but I'm unsure it's possible (good) or not.

**Reply by galeo** [reply]
@yakomaxa writes：

> Sorry for the delayed response, I just noticed today that I was mentioned weeks ago.
> 
> I'm really interested in your work. And I'm impressed that you reproduced it from my really rough descriptions. I feel sorry that I haven't opened the code.

Thank you so much for your message! No need to worry about the delay — I completely understand, and I really appreciate your exploration and the way you’ve shared your approach. It’s been very encouraging and inspiring.


> I have some questions; especially, what version of PyMOL did you use as the base code? I made mine using 2.x, but I believe today's PyMOL is 3.x. If it's based on 3.x, what happened to the GUI stuff when you ported it to WASM?
>
> I hacked PyMOL in a way that lacks proper git history (I mean I copied code, modified it, and pushed it to my private repo). I know this isn’t a good practice, and I wanted to restart clean from the official pymol-open-source repo before making it public. But I accidentally got really busy, and it’s been hard to find the time.

For my part, I worked from PyMOL 3.x (the master branch of official repository) when doing the WASM port. The git history is preserved, though I have to admit some of my changes are not very tidy — a few were made quite casually just to get the build to pass.

Regarding the GUI side, I think it’s pretty much the same as what you described before. I used the *gl4es* library and enabled the `PURE_OPENGL_ES_2` macro during compilation.

> I believe the core C++ code of PyMOL didn’t change much between 2.x and 3.x, so I think I could redo my work for 3.x.

Yes, that’s right — the core C++ code hasn’t changed much. I’m really looking forward to seeing your work on 3.x!

> Sorry for being vague, but I’d really appreciate any feedback. Thanks a lot for your attention!
> 
> If you are not uncomfortable, I’m willing to discuss more about this, including how to handle this situation.

I really appreciate the work you’ve done and the way you’ve shared it. So no need to apologize and I'm sorry for keeping this reply short. I’d be very happy to continue the discussion and exchange more details with you. 

> Added (after posted): I believe pymol-open-source devs has no responsibility for my Wasm version. I will be so happy if we could integrate it into official codebase, but I'm unsure it's possible (good) or not.

I agree. Still, I’d really love to see it integrated into the official codebase someday, and I’m curious what the maintainers’ perspective on that might be.

Thanks again for your inspiring work and for reaching out!

**Reply by yakomaxa** [reply]
Hi @galeo 

Thank you for your response! I really appreciate your efforts on another WASM port. It's impressive and encouraging to see it can be redone; it might sound like I'm joking, but I still feel a bit surreal to see PyMOL on the browser. It's excellent to hear that it's reproducible.

If you expect no conflicts between yours and ours, could I invite you to the private repository? I think you and we (I have another guy in that repo) will benefit, even just by reviewing something, while I expect you might find almost nothing technically new; I think you passed the most critical part.

I can invite @JarrettSJohnson too. 

I think my code remains less tidy and is almost frozen in practice, but this will help me be focused on this again!

**Reply by galeo** [reply]
@yakomaxa writes:

> If you expect no conflicts between yours and ours, could I invite you to the private repository? I think you and we (I have another guy in that repo) will benefit, even just by reviewing something, while I expect you might find almost nothing technically new; I think you passed the most critical part.

Thank you for the kind invitation! I’d be glad to join the private repo and help out wherever I can—even if it’s just reviewing things or giving small suggestions. It'll be nice to collaborate and check things together. Looking forward to it!

**Reply by yakomaxa** [reply]
@galeo 
Glad to hear that! I’ll get back to you in weeks or so. Things are getting a bit busy as I prepare for the next semester. Please hang tight! I’ll send you an invite and start thinking about how to reproduce it in 3.x code, extracting the modifications I made to create the patches.

**Comment by JarrettSJohnson** [comment]
> I agree. Still, I’d really love to see it integrated into the official codebase someday, and I’m curious what the maintainers’ perspective on that might be.

I've forwarded the comments to others regarding this, and there's ongoing discussions. But with PyMOL features planned in the next year or so, a web version of PyMOL in its current form is not ideal for us, so we may reconsider when those have landed.

**Reply by yakomaxa** [reply]
@JarrettSJohnson and also @galeo 

Thank you for your perspective. The WASM port has essentially the core C++/Python part only, and today's PyMOL (original) GUI part relies intensively on Python (PyQT) implementation. Discrepancy between WASM port and the original would come from this point.  

I tried to port the PyQt part (I imagined it will be superb if we could port external GUI too), but it was not possible for me. So, if we try to make a WASM version richer, I imagine something like "PyMOL binary wrapped by JavaScript/TypeScript," and it will be quite a different project than the original PyMOL's development trajectory which devs intend.

The PyMOL + JS/TS direction seems to be promising and even sounds hilarious, but I expect it might be misaligned with the dev's skill set and the direction of the PyMOL codebase. If so, I worry it could be a burden for the main codebase to have WASM/JS/TS parts in it.  WASM part + build scripts or so might be acceptable, but it just generates a GUI-less whl.

Hi, @hikarino-my, I would appreciate it if you could share your opinion about the points above.

**Reply by galeo** [reply]
@JarrettSJohnson writes:

> I've forwarded the comments to others regarding this, and there's ongoing discussions. But with PyMOL features planned in the next year or so, a web version of PyMOL in its current form is not ideal for us, so we may reconsider when those have landed.

Thanks you for the update. I’ll keep an eye out for any good news down the road, and hopefully can be revisited once the new features are in place.

**Reply by hikarino-my** [reply]
Since my workplace changed this year, I’ve been away from this project.
Recently, I’ve finally managed to make some time, so I thought I’d take another look at the current state. Also I’ve been in the mood to do something creative.
As for the GUI, my skill set leans more toward the web side, so I think it might be interesting if I could implement it with JavaScript or HTML, though it would probably end up being a complete reinvention of the wheel.
Since the project’s background has completely faded from my memory, I plan to reply sometime soon.

**Reply by yakomaxa** [reply]
@hikarino-my 
Thank you all! I'm also getting a bit busy, so no need to hurry. I’ll be responsive in weeks or so.

**Comment by pslacerda1** [comment]
I think this port is a good strategy as web tools are taking a lot market share. Recently I evaluated Molstar by reading their documentation, a proper selection engine is missing yet but otherwise may pretty much replace most visualization needs and maybe developers needs in near future also.

**Comment by labriata** [comment]
@yakomaxa 's pymol-wasm is one of the best things out there!
